From ffabfcf2abe3d5efcce0206363c4638b82d332a6 Mon Sep 17 00:00:00 2001
From: Prasanth Jayachandran <j.prasanth.j@gmail.com>
Date: Thu, 17 Dec 2015 13:38:57 -0600
Subject: [PATCH 529/587] CLOUDERA-BUILD: CDH-37763: HIVE-12684: NPE in stats
 annotation when all values in decimal column are
 NULLs (Prasanth Jayachandran reviewed by Pengcheng
 Xiong)

This appends a fix that is included in the backport on cdh5-1.1.0_dev and cdh5-1.1.0_5.7.0_dev.

Conflicts:
	ql/src/test/results/clientpositive/decimal_stats.q.out

Change-Id: Iaaa0186251d0215ee570e73a2bf4f1d8bb34abaa
---
 .../metadata/formatting/MetaDataFormatUtils.java   |   11 ++++++++++-
 .../results/clientpositive/decimal_stats.q.out     |    3 +--
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/formatting/MetaDataFormatUtils.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/formatting/MetaDataFormatUtils.java
index 1662696..534f90b 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/formatting/MetaDataFormatUtils.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/formatting/MetaDataFormatUtils.java
@@ -145,7 +145,16 @@ private static void formatAllFields(StringBuilder tableInfo,
   }
 
   private static String convertToString(Decimal val) {
-    return HiveDecimal.create(new BigInteger(val.getUnscaled()), val.getScale()).toString();
+    if (val == null) {
+      return "";
+    }
+
+    HiveDecimal result = HiveDecimal.create(new BigInteger(val.getUnscaled()), val.getScale());
+    if (result != null) {
+      return result.toString();
+    } else {
+      return "";
+    }
   }
 
   private static ColumnStatisticsObj getColumnStatisticsObject(String colName,
diff --git a/ql/src/test/results/clientpositive/decimal_stats.q.out b/ql/src/test/results/clientpositive/decimal_stats.q.out
index dabf7f8..cf0d242 100644
--- a/ql/src/test/results/clientpositive/decimal_stats.q.out
+++ b/ql/src/test/results/clientpositive/decimal_stats.q.out
@@ -63,7 +63,7 @@ STAGE PLANS:
       Map Operator Tree:
           TableScan
             alias: decimal_1
-            Statistics: Num rows: 500 Data size: 112000 Basic stats: COMPLETE Column stats: COMPLETE
+            Statistics: Num rows: 500 Data size: 5000 Basic stats: COMPLETE Column stats: COMPLETE
             Select Operator
               expressions: t (type: decimal(4,2)), u (type: decimal(5,0)), v (type: decimal(10,0))
               outputColumnNames: _col0, _col1, _col2
@@ -72,7 +72,6 @@ STAGE PLANS:
                 key expressions: 1 (type: int)
                 sort order: +
                 Statistics: Num rows: 500 Data size: 112000 Basic stats: COMPLETE Column stats: COMPLETE
-                TopN Hash Memory Usage: 0.1
                 value expressions: _col0 (type: decimal(4,2)), _col1 (type: decimal(5,0)), _col2 (type: decimal(10,0))
       Reduce Operator Tree:
         Select Operator
-- 
1.7.9.5

